src/util/bson/bson.cpp

```diff
@@ -443,641 +444,196 @@ std::ostream& operator<<(std::ostream& out, const Bson& b)
-                m_marks.top().pop_back();
-                m_marks.top().push_back(instruction.key, MongoTimestamp(val, ts.increment()));
-
-                // pop vestigal teimstamp instruction
-                m_instructions.pop();
-                m_instructions.push({State::Skip});
-                m_instructions.push({State::Skip});
-            } else {
-                m_marks.top().push_back(instruction.key, MongoTimestamp(val, 0));
-                instruction.type = State::Timestamp;
-            }
-            break;
-        default:
-            m_marks.top().push_back(instruction.key, static_cast<int64_t>(val));
-            break;
+namespace {
+// This implements just enough of the map API to support nlohmann's DOM apis that we use.
+template <typename K, typename V, typename ...Ignored>
+struct LinearMap {
+    using key_type = K;
+    using mapped_type = V;
+    using value_type = std::pair<K, V>;
+    using storage_type = std::vector<value_type>;
+    using iterator = typename storage_type::iterator;
+    using const_iterator = typename storage_type::const_iterator;
+
+    auto begin() { return _elems.begin(); }
+    auto begin() const { return _elems.begin(); }
+    auto end() { return _elems.end(); }
+    auto end() const { return _elems.end(); }
+    auto size() const { return _elems.size(); }
+    auto max_size() const { return _elems.max_size(); }
+    auto clear() { return _elems.clear(); }
+    V& operator[](const K& k) {
+        // assume this is only used for adding a new element.
+        return _elems.emplace_back(k, V()).second;
     }
-    return true;
-}
 
-bool Parser::number_float(number_float_t val, const string_t&) {
-    if (m_instructions.size() && m_instructions.top().type != State::StartArray) {
-        auto instruction = m_instructions.top();
-        m_instructions.pop();
-        m_marks.top().push_back(instruction.key, static_cast<double>(val));
+    template <typename... Args>
+    std::pair<iterator, bool> emplace(Args&&... args) {
+        // assume this is only used for adding a new element.
+        _elems.emplace_back(std::forward<Args>(args)...);
+        return {--_elems.end(), true};
     }
-    // if there have been no previous instructions
-    // this is considered fragmented JSON, which is allowed.
-    // this means we have received a raw double
-    else {
-        m_marks.top().push_back("", static_cast<double>(val));
-    }
-
-    return true;
-}
 
-/*!
- @brief a string was read
- @param[in] val  string value
- @return whether parsing should proceed
- @note It is safe to move the passed string.
- */
-bool Parser::string(string_t& val) {
-    if (!m_instructions.size()) {
-        m_marks.top().push_back("", std::string(val.begin(), val.end()));
-        return false;
+    iterator erase(iterator) {
+        // This is only used when mutating the DOM which we don't do.
+        REALM_ASSERT_RELEASE(!"LinearMap::erase() should never be called");
```

Comment:
`REALM_TERMINATE()` is a more straightforward way to do this.

Reply:
done